domain/tutoringModel/ga.py

```diff
@@ -155,25 +166,61 @@ def calculate_learning_path_ga(self, learning_style):
             best_score = fitness[0]
 
             # update population
-            new_pop = self.population[best_index].copy()
-            self.population = new_pop.copy()
+            new_pop = self.population[best_index]
+            self.population = new_pop
             best_sample = new_pop[0]
 
-            if best_score < best_total_score and i < self.n_generation:
+            if best_score < best_total_score:
                 best_total_score = best_score
                 self.best_population = best_sample
+                np.insert(best_sample, 0, 0)
+
+            i += 1
 
             # evolution
             self.evolve(fitness, best_sample)
 
-        ga_path = np.array(self.learning_elements)
-        population = self.valide_population()
-        idx = population[0]
-        result_ga_lp = ga_path[idx]
-
-        learning_path = utils.get_learning_path_as_str(result_ga_lp)
-
-        return learning_path
+            if i > self.max_generation:
+                best_sample = self.seach_learning_elements()
+
+            valid = self.valide_elementen(
+                self.learning_elements[np.insert(best_sample, 0, 0)]
+            )
+
+        # add the first element to the population
+        if not self.first_is_present:
+            best_sample = np.insert(best_sample, 0, 0)
+
+        # sort the learning elements
+        self.learning_elements = self.learning_elements[best_sample]
+        self.le_coordinate = self.le_coordinate[best_sample]
+
+        learning_path_as_str = ", ".join(self.learning_elements)
+        if len(self.learning_elements) == 1:
+            learning_path_as_str = self.learning_elements[0] + ", "
+
+        return learning_path_as_str
+
+    def valide_elementen(self, learning_elements):
+        result = learning_elements
+        if "KÜ" in result and result[0] != "KÜ":
+            return True
+        if "EK" in result and not (result[0] == "EK" or result[1] == "EK"):
```

Comment:
and (result[0] != "EK" or result[1] != "EK"). Damit ist es identisch zu KÜ und LZ aufgebaut und ist leichter lesbar

Reply:
Done